# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/workers/remoteworkers/RemoteWorkerParent.cpp
# Commit: b55aacf910e7
# Full Hash: b55aacf910e704508a5820c09b25411142ccf7cb
# Author: Nika Layzell <nika@thelayzells.com>
# Date: 2024-06-25 09:36:02
# Regressor Bug: 1728331
# File Overlap Count: 1
# Description:
#   Bug 1728331 - Part 4: Make ContentParent KeepAlives explicit with RAII references, r=smaug,dom-worker-reviewers,asuth
#   
#   This is a fairly significant patch, however it would be difficult to break it
#   down into smaller patches:
#   
# ==============================================================================

diff -r 3dd07ed49638 -r b55aacf910e7 dom/workers/remoteworkers/RemoteWorkerParent.cpp
--- a/dom/workers/remoteworkers/RemoteWorkerParent.cpp	Mon Jun 24 23:19:28 2024 +0000
+++ b/dom/workers/remoteworkers/RemoteWorkerParent.cpp	Mon Jun 24 23:19:28 2024 +0000
@@ -10,9 +10,7 @@
 #include "mozilla/dom/ContentParent.h"
 #include "mozilla/dom/PFetchEventOpProxyParent.h"
 #include "mozilla/ipc/BackgroundParent.h"
-#include "mozilla/SchedulerGroup.h"
 #include "mozilla/Unused.h"
-#include "nsProxyRelease.h"
 
 namespace mozilla {
 
@@ -20,34 +18,9 @@
 
 namespace dom {
 
-namespace {
-
-class UnregisterActorRunnable final : public Runnable {
- public:
-  explicit UnregisterActorRunnable(
-      already_AddRefed<ThreadsafeContentParentHandle> aParent)
-      : Runnable("UnregisterActorRunnable"), mContentHandle(aParent) {
-    AssertIsOnBackgroundThread();
-  }
-
-  NS_IMETHOD
-  Run() override {
-    AssertIsOnMainThread();
-    if (RefPtr<ContentParent> contentParent =
-            mContentHandle->GetContentParent()) {
-      contentParent->UnregisterRemoveWorkerActor();
-    }
-
-    return NS_OK;
-  }
-
- private:
-  RefPtr<ThreadsafeContentParentHandle> mContentHandle;
-};
-
-}  // namespace
-
-RemoteWorkerParent::RemoteWorkerParent() {
+RemoteWorkerParent::RemoteWorkerParent(
+    UniqueThreadsafeContentParentKeepAlive aKeepAlive)
+    : mContentParentKeepAlive(std::move(aKeepAlive)) {
   AssertIsOnBackgroundThread();
   MOZ_ASSERT(XRE_IsParentProcess());
 }
@@ -62,21 +35,6 @@
       PRemoteWorkerParent::Manager());
 }
 
-void RemoteWorkerParent::Initialize(bool aAlreadyRegistered) {
-  RefPtr<ThreadsafeContentParentHandle> parent =
-      Manager()->GetContentParentHandle();
-
-  // Parent is null if the child actor runs on the parent process.
-  if (parent) {
-    if (!aAlreadyRegistered) {
-      parent->RegisterRemoteWorkerActor();
-    }
-
-    NS_ReleaseOnMainThread("RemoteWorkerParent::Initialize ContentParent",
-                           parent.forget());
-  }
-}
-
 already_AddRefed<PFetchEventOpProxyParent>
 RemoteWorkerParent::AllocPFetchEventOpProxyParent(
     const ParentToChildServiceWorkerFetchEventOpArgs& aArgs) {
@@ -88,15 +46,7 @@
   AssertIsOnBackgroundThread();
   MOZ_ASSERT(XRE_IsParentProcess());
 
-  RefPtr<ThreadsafeContentParentHandle> parent =
-      Manager()->GetContentParentHandle();
-
-  // Parent is null if the child actor runs on the parent process.
-  if (parent) {
-    RefPtr<UnregisterActorRunnable> r =
-        new UnregisterActorRunnable(parent.forget());
-    SchedulerGroup::Dispatch(r.forget());
-  }
+  mContentParentKeepAlive = nullptr;
 
   if (mController) {
     mController->NoteDeadWorkerActor();